src/catchup/CatchupWork.cpp

```diff
@@ -346,113 +351,85 @@ CatchupWork::onSuccess()
 namespace
 {
 
-// compute first checkpoint that is not 100% finished
-// if lastClosedLedger is not last ledger in checkpoint, then first not finished
-// checkpoint is checkpoint containing lastClosedLedger
-// if lastClosedLedger is last ledger in checkpoint, then first not finished
-// checkpoint is next checkpoint (with 0 ledgers applied)
-uint32_t
-firstNotFinishedCheckpoint(uint32_t lastClosedLedger,
-                           HistoryManager const& historyManager)
-{
-    auto result = historyManager.checkpointContainingLedger(lastClosedLedger);
-    if (lastClosedLedger < result)
-    {
-        return result;
-    }
-    else
-    {
-        return result + historyManager.getCheckpointFrequency();
-    }
-}
-
 // return first ledger that should be applied so at least count history entries
 // are stored in database (count 0 is changed to 1, because even doing only
 // bucket apply gives one entry)
 //
 // if it is impossible, returns smallest possible ledger number -
 // LedgerManager::GENESIS_LEDGER_SEQ
 uint32_t
-firstNeededLedger(CatchupConfiguration const& configuration)
+firstNeededLedger(uint32_t toLedger, uint32_t count)
 {
-    auto neededCount = std::max(configuration.count(), 1u);
-    return configuration.toLedger() > neededCount
-               ? configuration.toLedger() - neededCount + 1
-               : LedgerManager::GENESIS_LEDGER_SEQ;
+    auto neededCount = std::max(count, 1u);
+    return toLedger > neededCount ? toLedger - neededCount + 1
+                                  : LedgerManager::GENESIS_LEDGER_SEQ + 1;
 }
 
-std::pair<bool, uint32_t>
-computeCatchupStart(uint32_t smallestLedgerToApply,
-                    uint32_t smallestBucketApplyCheckpoint,
+uint32_t
+computeCatchupStart(uint32_t lastClosedLedger,
+                    CatchupConfiguration const& configuration,
                     HistoryManager const& historyManager)
 {
-    // checkpoint that contains smallestLedgerToApply - it is first one than
-    // can be applied, it is always greater than LCL
-    auto smallestCheckpointToApply =
-        historyManager.checkpointContainingLedger(smallestLedgerToApply);
-
-    // if first ledger that should be applied is on checkpoint boundary then
-    // we do an bucket-apply
-    if (smallestLedgerToApply == smallestCheckpointToApply)
+    if (lastClosedLedger == 0)
     {
-        return std::make_pair(true, smallestCheckpointToApply);
+        throw std::invalid_argument{"lastClosedLedger == 0"};
     }
 
-    // we are before first checkpoint - applying buckets is not possible, so
-    // just apply transactions
-    if (smallestLedgerToApply < historyManager.getCheckpointFrequency() - 1)
+    if (configuration.toLedger() <= lastClosedLedger)
     {
-        return std::make_pair(false, smallestLedgerToApply);
+        throw std::invalid_argument{
+            "configuration.toLedger() <= lastClosedLedger"};
```

Comment:
curious about this error: do we expect end users to see this or not?
If so, we should have a better error message than this.
I think we expect this check to be done upstream so probably not the case.

Reply:
End user should not check this error, but I think that we are missing one case here. When user starts catchup to current ledger, but local history is newer than remote history (in case of testnet error).

I'll add new check, but otherwise this should not be called at all (which is why I preferred this to be an assert).

Reply:
Correction: this check exists, but is in different place than expected.